pkg/controller/knativeserving/openshift/openshift.go

```diff
@@ -581,3 +598,121 @@ func installNetworkPolicies(instance *servingv1alpha1.KnativeServing) error {
 	}
 	return nil
 }
+
+func exposeMetricToServiceMonitor(instance *servingv1alpha1.KnativeServing) error {
+	// getOperatorNamespace return namespace where knative-serving-operator has been installed
+	namespace, err := getOperatorNamespace()
+	if err != nil {
+		log.Info("no namespace defined, skipping ServiceMonitor installation for the operator")
+		return nil
+	}
+
+	log.Info("exposing metric for installed knative version")
+	knativeVersion.WithLabelValues(version.Version, namespace).Set(float64(time.Now().Unix()))
+
+	// Get service knative-serving-operator in order to create service monitor
+	svcDetail, err := getOperatorService(namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info(fmt.Sprintf("service knative-serving-operator does not exist in %s namespace", namespace))
+			return nil
+		}
+		return err
+	}
+
+	// Add label openshift.io/cluster-monitoring to namespace
+	if err = addMonitoringLabelToNamespace(namespace); err != nil {
+		return err
+	}
+
+	// CreateServiceMonitors creates ServiceMonitors objects based on Service objects.
+	if _, err := metric.CreateServiceMonitors(cfg, namespace, []*v1.Service{svcDetail}); err != nil {
+		if errors.IsAlreadyExists(err) {
+			log.Info(fmt.Sprintf("ServiceMonitor already exist for %s service", svcDetail.Name))
+			return nil
+		}
+		log.Error(err, "failed to create ServiceMonitor")
+		return err
+	}
+	log.Info("successfully created ServiceMonitor")
+	if err = createRoleForServiceMonitor(namespace); err != nil {
+		return err
+	}
+	if err = createRoleBindingForServiceMonitor(namespace); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getOperatorService(ns string) (*v1.Service, error) {
+	svc := &v1.Service{}
+	err := api.Get(context.TODO(), types.NamespacedName{Name: "knative-serving-operator", Namespace: ns}, svc)
+	return svc, err
+}
+
+func getOperatorNamespace() (string, error) {
+	ns, found := os.LookupEnv(knativeServingInstalledNamespace)
+	if !found {
+		return "", fmt.Errorf("%s must be set", knativeServingInstalledNamespace)
+	}
+	return ns, nil
+}
+
+func addMonitoringLabelToNamespace(namespace string) error {
+	ns := &v1.Namespace{}
+	if err := api.Get(context.TODO(), client.ObjectKey{Name: namespace}, ns); err != nil {
+		return err
+	}
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
+	}
+	ns.Labels[monitoringLabel] = "true"
+	if err := api.Update(context.TODO(), ns); err != nil {
+		log.Error(err, fmt.Sprintf("could not add label %q to namespace %q", monitoringLabel, namespace))
+		return err
+	}
+	return nil
+}
+
+func createRoleForServiceMonitor(ns string) error {
```

Comment:
These two functions need to be idempotent to correctly pass the reconcile flow.

That usually goes like this:

```golang
obj, err := api.Get(...)
if errors.IsNotFound(err) {
  obj := constructObj()
  return api.Create(obj)
} else if err != nil {
  return err
} else {
  obj := constructObj()
  return api.Update(obj)
}
```

Have you thought about creating a similar manifest to that we already have in `/deploy/resources/monitoring` and run a transformer over it that adds the namespace at runtime? That'd make this a tad bit easier as that handles all the edge-cases for you I believe.

Reply:
@markusthoemmes Thank you 
I have modified according to the comment.